### ISSUE TITLE
Integrating disqus in a mobile app

### ISSUE THREAD

**Issue by yiannistri** [disqus/DISQUS-API-Recipes#22]
Hi,

I'm following your [guide](https://help.disqus.com/customer/portal/articles/472096) to display disqus comments on my Android app. It works fine for the most part as comments are being displayed nicely. However, I can't get the login flow to work whether it's via email or through any of the social media. The result is that my users can see the comments but cannot login or add any new comments.Anyone following this guide can easily reproduce this behaviour. I'm suspecting that the popup window that you now display on clicking Login doesn't play well with embedded webview.
Any suggestions on how to solve this?

**Comment by jasonatt** [comment]
Same issue here... Any feedback please?

**Comment by dmatt** [comment]
Hi @yiannistri and @jasonatt. Thank you for reporting this issues. There is indeed a problem with the login flow, which we became aware of recently. We have a fix that we're working on addressing in that documentation soon, likely this week. Apologies for the outdated documentation.

**Comment by jasonatt** [comment]
Hi, that was fast :) 

Will you notify us via this thread or shall I subscribe somewhere?

Cheers

**Comment by dmatt** [comment]
Yes :). I'll be sure to mark this issue as closed once we fix the
documentation. Subscribing to this issue should be good.

**Comment by jasonatt** [comment]
Cool Thanks

**Comment by RamizSami** [comment]
Hello @dmatt! I've to get Disqus integrated in my android application and for that I'm waiting for the new documentation. Any updates?

**Comment by dmatt** [comment]
We'll have some updated instructions for android very soon (most likely
this week), sorry for the wait. I'll be sure to update this Issue once we
have more to share.

**Comment by therealmax** [comment]
Having the same issue, not able to comment after navigating back to the comment thread url when a successful login is detected.

**Comment by RamizSami** [comment]
@dmatt any updates?

**Comment by dmatt** [comment]
Yes, the mobile instructions are still a work in progress. Our developers
have an updated Windows UWP example here
https://github.com/ryanvalentin/disqus-install-examples/tree/master/windows-uwp.
We will be adding iOS and android soon but no ETA yet.

**Comment by therealmax** [comment]
I found that the issue doesn't exist in android 4.4.2, but android 5 and above still have the issue.

**Comment by taml** [comment]
@dmatt Hello, Is there going to be Javascript documentation for this? I've had a look through the windows documentation and each time I'm trying to sign in I'm not seeing any of these url endings "#!auth%3Acancel", "#!auth%3Asuccess", #!auth%3Afail" to be able to redirect anywhere, it's just sticking on the auth start one, but if I restart the app I will be signed in. I'd love to get this working asap as I've spent a good few days on it now.

**Comment by jasonatt** [comment]
Hi, I can confirm taml issue. I have a cordova build of the app and on ios I can see the ending urls but on android does nothing so cannot redirect.

**Comment by taml** [comment]
Hey @jasonatt have you managed to get the redirects working on iOS? I'm building a Ionic / Cordova App too! It would be good to compare code to see how you've implemented Disqus as I'm getting Disqus to show the right comments for the right posts 98% of the time ( for some reason it's not bringing the right ones in on two posts) but haven't managed to get anywhere at all with the redirects!

**Comment by jasonatt** [comment]
Hi @taml,

I am using the in app browser and listen to when the page finishes loading to detect the url and redirect based on that. Perhaps @dmatt can assist as I think the paths are different on android...

     $scope.disqusUrl = "Your Url of the starting page"
      
    $scope.disqusUrlCall = function(url) {
     var ref = cordova.InAppBrowser.open(encodeURI($scope.disqusPushUrl), '_blank', 'location=no', clearsessioncache="yes", clearcache="yes");
     ref.addEventListener('loadstop', checkUrl);
    };

    function checkUrl(params) {
     //console.log(params.url);
     if (params.url.indexOf("disqus.com/next/login-success") > -1 || params.url.indexOf("disqus.com/_ax/facebook/complete") > -1 || params.url.indexOf("disqus.com/_ax/google/complete") > -1 || params.url.indexOf("disqus.com/_ax/twitter/complete") > -1) {
     //alert("Redirect from Disqus");
     cordova.InAppBrowser.open(encodeURI($scope.disqusUrl), '_blank', 'location=no');
    } 
    }

**Comment by taml** [comment]
Hi @jasonatt Thanks for sharing your code. Where are you putting that code? At the moment I have this in my controllers.js file
`$scope.showComments = function (postid, posttitle, posturl) {

    var disqus_identifier = postid;
    var disqus_title = posttitle;
    var disqus_url = posturl;
    var url = "views/app/disqus.html?";
    $scope.disqusurl = url + "shortname=mysiteshortname&url=" + encodeURIComponent(disqus_url) +
    "&title=" + encodeURIComponent(disqus_title) + "&identifier=" + encodeURIComponent(disqus_identifier);
    $scope.disqusurl = $sce.trustAsResourceUrl($scope.disqusurl);
};
`
I'm also using a directive linked to the Cordova in app browser so I can target a specific link to open in the in app browser by calling it in a <a href... dynamic-anchor> </a>

**Comment by jasonatt** [comment]
Hi,

Its in a controller.
I have a button for the user to open the in app browser with disqus by calling the disqusUrlCall function. 

However my url is located from the server as I have the disqus script running from there so I basically send the parameters to the server and show the comments page via inapp browser. its the only way i found to work properly.

**Comment by jasonatt** [comment]
@dmatt any updates? 

Thanks

**Comment by ryanvalentin** [comment]
So it sounds like we would also need instructions on how to put Disqus on a Ionic/Cordova app as well. I can look into that, but I'm not exactly sure what the pitfalls are off the top of my head.

**Comment by jasonatt** [comment]
The main issue I found it that after the user logs in I cannot redirect them on Android. It works on ios however.

**Comment by taml** [comment]
@ryanvalentin Instructions for Ionic / Cordova would be so helpful!

**Comment by georgekosmidis** [comment]
Hey @ryanvalentin  , any news on how to add disqus on a Cordova app?

Are there any plans for this?
I 'd like to know cause I can wait and release the app with disqus commenting system, if it's in your immediate plans...

Thanks!

**Comment by MaxTimothy** [comment]
@taml @georgekosmidis I have disqus working (login/out/comment) with a purely javascript implementation on android. What problems are you all having, I might be able to help out.

**Comment by jrossfeld** [comment]
For those on iOS that found this broken, the key url to watch out for is "https://disqus.com/next/login/?forum=myforumname#!auth%3Astart" If you were reloading the comments after login success and waiting for "https://disqus.com/next/login/?forum=myforumname" That won't work anymore. I had to specifically make sure the auth%3Astart one went through without reloading the comments again for it to work.

**Comment by Brle88** [comment]
Has anyone tried to implement this to ionic 2 / angular 2?

**Comment by pablogonzales** [comment]
@Brle88  http://stackoverflow.com/questions/36102556/is-this-the-right-way-to-embed-disqus-in-angular2-component maybe this can help you

**Comment by correasebastian** [comment]
@MaxTimothy can you help me with an example(repo or code) , how you are handling the login with disqus in a hybrid app

**Comment by correasebastian** [comment]
@dmatt any updates?

Thanks

**Comment by MaxTimothy** [comment]
@correasebastian Show me some code so I can see what you've tried/doing and I'll see if I can help, but with out it I have no way of knowing what's going on in your program.

**Comment by correasebastian** [comment]
@MaxTimothy thanks for you answer, 

Im using
Cordova CLI: 6.4.0
Gulp version:  CLI version 3.9.1
<plugin name="cordova-plugin-inappbrowser" spec="~1.6.1" />
OS: Mac OS X El Capitan
Node Version: v6.7.0

**testing on android 5.1.0**
this is my code  

```
var disqusUrl = 'https://kyky.localtunnel.me/?shortname=ufwtest&url=https://ufw-app-staging.herokuapp.com/&title=third';
var inAppCall;

function disqusUrlCall() {
  var options = "location=yes,clearsessioncache=yes,clearcache=yes"
  inAppCall = cordova.InAppBrowser.open(encodeURI(disqusUrl), '_blank', options);
  inAppCall.addEventListener('loadstop', checkUrl);
};

function checkUrl(params) {
  //console.log(params.url);
  if (params.url.indexOf("disqus.com/next/login-success") > -1 || params.url.indexOf("disqus.com/_ax/facebook/complete") > -1 || params.url.indexOf("disqus.com/_ax/google/complete") > -1 || params.url.indexOf("disqus.com/_ax/twitter/complete") > -1) {


    // inAppCall.removeEventListener('loadstop', checkUrl);
    /*
    // first attemp: this does not work, after openning  ask for login  again
     cordova.InAppBrowser.open(encodeURI(disqusUrl), '_blank', 'location=no');
    */


    /*
    // second  attemp: this does not work, after redirecting  ask for login  again
      var script = 'window.location="' + disqusUrl + '"';
      debugger;
      inAppCall.executeScript({
        code: script
      }, function(params) {
        console.log('script execute successfully', params)
      })
    */

  }
}

disqusUrlCall()
```

this is the path:

1. Open inappbrowser and load the disqus embed

2. i click login, and succesfully logged in

3.  redirect to the disqus forum ( but is not logged in)

any help will be appreciate

**Comment by MaxTimothy** [comment]
@correasebastian so I think your inappbrowser needs to be accepting 3rd party cookies. After lollipop that's no long done by default.

**Comment by correasebastian** [comment]
@MaxTimothy  thanks again for you answer, but can you point me in the right direction, i cant find how to achieve that in the documentation, i only see this pull request open https://github.com/apache/cordova-plugin-inappbrowser/pull/198/commits

**Comment by MaxTimothy** [comment]
@correasebastian Unfortunately I've never developed with cordova, but I would think you could add it as an option.

**Comment by ayyfish** [comment]
Just wondering if there are any updates on this. In the Android InAppBrowser I am able to login successfully and redirect back to the discussion page. On the discussion page, however, it shows that I'm not logged in. When I try to login again I get an error page saying "There was an error submitting the form." On that page it says try repeating the action on disqus.com. When navigating to disqus.com from a link on the error page, it shows that I am logged in correctly. So login isn't working only for the discussion thread. On iOS login is working fine.

**Comment by gameleon-dev** [comment]
Another issue that popped up around mid-May is that the google OAuth request system doesnt allow web requests anymore from a webview like UIWebView and the Android Webview, and suggests to use  the native libraries for Google Sign-In etc instead, thereby breaking the Google login for the Disqus native embed as well.

![afbeelding](https://user-images.githubusercontent.com/5390147/27468446-2b7328cc-57ea-11e7-8d0f-38a4c472196d.png)

**Comment by boxOfAces** [comment]
@dmatt Has there been any updates on this officially? Or do we have to use the "listening for url changes" hack?

**Comment by MaxTimothy** [comment]
@gameleon-dev change your webview's user agent using setUserAgentString, google auth uses this to tell if the user is signing in through a webview(which is no longer allowed). This is how you trick it.

**Comment by Bishal18** [comment]
@MaxTimothy i tried changing the user agent in the InAppBrowser with the native user agenyt set and although it's changing the user agent but still it's a 403 error i am facing. Saw somewhere that if the webview is using chromium 67 there is a isuue with multiple windows where the user agent is set for the first window but it's fixed in version 68. Any help is appreciated.

**Comment by MaxTimothy** [comment]
@Bishal18 changing the user agent works for me regardless of the chromium version, what are you setting your user agent to?

**Comment by Alloy990** [comment]
[](url) DISQUS

**Comment by rosaWetteloos** [comment]
Any updates on this issue? I honestly think updating the user agent is a valid solution. I've ran into some issues with other third party services that require me to keep the user agent as it is.

**Comment by premiumappapk332** [comment]
Download android apps, games, themes and live wallpapers direct APK for all android smartphones, tablets and other devices from <a href="http://premiumappapk.com/Main/show/46">Apps Apk</a>

**Comment by hristod** [comment]
Hello all, is there any update on this issue? We're struggling to implement a web view in our Android app and have the same issue. Is there an alternative way of doing that? Any help is appreciated! Cheers!

**Comment by gvsrepins** [comment]
@hristod maybe this post can help you
https://help.disqus.com/en/articles/1717165-javascript-embed-in-native-apps

I'm also starting to implement this solution on a Ionic v1.x hybrid app.

Let me know if you have success. 

Thanks!

**Comment by Sahejmaharjan1** [comment]
Hi, I am also facing the same issue in react native with login flow... I get to a loading page after login.. it doesn't redirect back to the app nor I am logged In.. Anyone solved this issue?

**Comment by parthkanani93** [comment]
> Hi, I am also facing the same issue in react native with login flow... I get to a loading page after login.. it doesn't redirect back to the app nor am I logged In.. Anyone solved this issue?

are you able to run an embedded web view on ios react-native? if yes then how?
  i used  https://github.com/isrmicha/react-native-disqus  library but it is not working on ios...

**Comment by mandeepk91** [comment]
> > Hi, I am also facing the same issue in react native with login flow... I get to a loading page after login.. it doesn't redirect back to the app nor am I logged In.. Anyone solved this issue?
> 
> are you able to run an embedded web view on ios react-native? if yes then how? i used https://github.com/isrmicha/react-native-disqus library but it is not working on ios...

Hi, I am trying to implement in ios but fails. u have any idea?
